### PR TITLE
Add pagination to agent history, serialize metadata to prevent OOM

### LINKED ISSUE #38: Limit the agent history
### What feature do you want to see added?

When an agent is heavily used by many jobs that are fast and run frequently the history can easily reach thousands of runs. This can lead to noticable delays when trying to open the details of a run to show all the matching steps (might also be caused by the YUI stuff).
By limiting the number of entries this can reduce load on the backend and speed up the UI.
Paging might also be a solution in case the ui should be fast but we need access to older entries.

### Upstream changes

_No response_

### Are you interested in contributing this feature?

_No response_

### PR DESCRIPTION
**Pull request by ChristianMa97**
Fixes #38 
#### **Summary**
This PR introduces optimizations and new features to the **Agent Build History Plugin**. The core enhancement is the shift from holding build execution data in memory to storing metadata as serialized objects, significantly reducing memory usage. Key features include:

1. **Memory Optimization**: 
   - Builds metadata are now added to a serialized list per node during synchronization and on new runs. 
   - Only the builds that need to be displayed are loaded into memory, ensuring efficient resource utilization.

2. **Pagination**:
   - Introduced pagination for node build history. Only the relevant set of builds for each page is loaded.
   - Users can navigate through the paginated build history with customizable entries per page.

3. **UI Enhancements**:
   - Added new UI elements in the build history table to support sorting and pagination.
   - The configuration page now allows users to define default settings for entries per page, sorting preferences and serialization directory.

#### **Key Changes**
- **Core Functionalities**:
    - **Serialized Build Execution Storage**: Builds are stored as serialized metadata per node, ensuring that large build histories do not consume excessive memory.
    - **Pagination and Sorting**: Added pagination and sorting capabilities to the node build history view. Users can navigate through pages of builds and sort by start time or build number.
  
- **File Management Enhancements**:
    - **Index File Management**: Added support for creating and managing index files per node, ensuring that build history is correctly tracked and accessible even after restarts or configuration changes.
    - **Job and Node Renaming**: Implemented logic to handle job and node renaming, ensuring that the build history is correctly updated.
    - **Build, Job and Node Deletion**: Implemented logic to hande deletions, ensuring that the build history is correctly updated.

#### **Technical Details**
- **AgentBuildHistory Class**:
    - Refactored to load build execution data from serialized files rather than storing it in memory.
    - Introduced methods for paginating and sorting build data.

- **BuildHistoryFileManager Class**:
    - Handles reading from and writing to index files, ensuring that only the relevant builds are loaded into memory during each page request.
    - Added methods for renaming jobs and nodes in index files, as well as for deleting serialized data when a job or node is deleted.

- **AgentBuildHistoryConfig Class**:
    - Added configuration options for defining the default storage directory, entries per page, default sorting column, and order.
    - Automatically creates and manages the storage directory based on the Jenkins environment variable `JENKINS_HOME` or a default path.

#### **UI/UX Changes**:
- **Pagination Controls**: 
    - Added pagination controls to the build history view with customizable page sizes.
- **Sorting Options**:
    - Users can sort builds by start time or build name and number, in ascending or descending order.

### Testing done
- Tested on a Jenkins instance running in Docker with two agents.
- Deployed and tested on two smaller production systems.
- Still planning to test on a larger system later.
- Automated tests cover:
    - Build serialization and deserialization.
    - Pagination and sorting of build history.
    - Job and node renaming, ensuring that the build history is properly updated.
    - Deletion of jobs, nodes, and builds from the serialized files.

-->

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/resources/io/jenkins/plugins/agent_build_history/AgentBuildHistory/index.jelly [hunk 1]
```diff
@@ -34,17 +34,76 @@
...
+      <j:set var="pageSize" value="${request.getParameter('pageSize') != null ? request.getParameter('pageSize') : it.entriesPerPage}"/>
+      <j:set var="totalPages" value="${it.totalPages}"/>
+      <j:set var="sortColumn" value="${request.getParameter('sortColumn') != null ? request.getParameter('sortColumn') : it.defaultSortColumn}"/>
+      <j:set var="sortOrder" value="${request.getParameter('sortOrder') != null ? request.getParameter('sortOrder') : it.defaultSortOrder}"/>
+      <div class="pagination-container">
+        <!-- Pagination Previous/Next Buttons -->
+        <nav aria-label="Page navigation" class="pagination">
+          <!-- Previous button -->
+          <j:if test="${page &gt; 1}">
+            <a class="page-link" href="${request.requestURI}?page=${page - 1}&amp;pageSize=${pageSize}&amp;sortColumn=${sortColumn}&amp;sortOrder=${sortOrder}" aria-label="Previous">
```
Comment (mawinter69) [inline src/main/resources/io/jenkins/plugins/agent_build_history/AgentBuildHistory/index.jelly:48]: Add `tooltip="${%Previous}"`

File: src/main/resources/io/jenkins/plugins/agent_build_history/AgentBuildHistory/index.jelly [hunk 2]
```diff
@@ -34,17 +34,76 @@
...
+          </j:if>
+
+          <!-- Page Input Field -->
+          <span class="page-label">Page</span>
+          <input type="number" class="form-control page-input" min="1" max="${totalPages}" value="${page}" onchange="window.location.href='${request.requestURI}?page=' + this.value + '&amp;pageSize=${pageSize}&amp;sortColumn=${sortColumn}&amp;sortOrder=${sortOrder}'"/>
+          <span class="page-label">of ${totalPages}</span>
+
+          <!-- Next button -->
+          <j:if test="${page &lt; totalPages}">
+            <a class="page-link" href="${request.requestURI}?page=${page + 1}&amp;pageSize=${pageSize}&amp;sortColumn=${sortColumn}&amp;sortOrder=${sortOrder}" aria-label="Next">
```
Comment (mawinter69) [inline src/main/resources/io/jenkins/plugins/agent_build_history/AgentBuildHistory/index.jelly:60]: Add `tooltip="${%Next}"`

File: src/main/java/io/jenkins/plugins/agent_build_history/AgentBuildHistoryConfig.java
Comment (mawinter69) [inline src/main/java/io/jenkins/plugins/agent_build_history/AgentBuildHistoryConfig.java:1]: I think it is not a good idea to have global defaults for entries per page, sort field and sort direction. This should be a per user setting. Maybe even set this in a cookie so when I change this it stays on every page. E.g. the icon size is also stored in a cookie.

File: src/main/java/io/jenkins/plugins/agent_build_history/AgentBuildHistoryConfig.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    private String defaultSortColumn = "startTime";
+    private String defaultSortOrder = "desc";
+
+    public AgentBuildHistoryConfig() {
+        load(); // Load the persisted configuration
+        ensureStorageDir();
+    }
+
+    private String getDefaultStorageDir() {
+        String jenkinsHome = System.getenv("JENKINS_HOME");
```
[the remark this answers is not in the document]
Comment (ChristianMa97) [reply src/main/java/io/jenkins/plugins/agent_build_history/AgentBuildHistoryConfig.java]: I changed it to use Jenkins.get().getRootDir() as suggested. The logic for changing the directory is already in place, so I'd prefer to keep it. The default should work fine for most users, but it's nice to have the option to adjust if needed. What do you think?

File: src/main/resources/io/jenkins/plugins/agent_build_history/agentBuildHistory.css [hunk 4]
```diff
@@ -90,3 +90,27 @@
...
+
+.abh-pagination {
+  display: flex;
+  align-items: center;
+}
+
+.abh-page-size-selector {
+  display: flex;
+  align-items: center;
+  gap: 10px;
```
Comment (mawinter69) [inline src/main/resources/io/jenkins/plugins/agent_build_history/agentBuildHistory.css:111]: reduce gap to `5px`

File: src/main/resources/io/jenkins/plugins/agent_build_history/agentBuildHistory.css [hunk 5]
```diff
@@ -90,3 +90,27 @@
   gap: 18px;
   padding: 0 15px;
 }
+
+.abh-pagination-container {
+  display: flex;
+  flex-direction: row;
+  align-items: center;
+  justify-content: flex-start;
+  margin-top: 20px;
```
Comment (mawinter69) [inline src/main/resources/io/jenkins/plugins/agent_build_history/agentBuildHistory.css:99]: the `margin-top` is not needed
instead add some minimal margin to the bottom so the inputs are a but separated from the table below

File: src/main/resources/io/jenkins/plugins/agent_build_history/agentBuildHistory.css [hunk 6]
```diff
@@ -90,3 +90,27 @@
...
+  gap: 20px;
+}
+
+.abh-pagination {
+  display: flex;
+  align-items: center;
```
Comment (mawinter69) [inline src/main/resources/io/jenkins/plugins/agent_build_history/agentBuildHistory.css:105]: add a `gap: 5px;`

File: src/main/resources/io/jenkins/plugins/agent_build_history/AgentBuildHistory/index.jelly [hunk 7]
```diff
@@ -34,19 +34,86 @@
...
+            <a class="jenkins-button" tooltip="${%Next}" href="${request.requestURI}?page=${page + 1}&amp;pageSize=${pageSize}&amp;sortColumn=${sortColumn}&amp;sortOrder=${sortOrder}" aria-label="Next">
+              <l:icon src="symbol-arrow-forward-outline plugin-ionicons-api"/>
+            </a>
+          </j:if>
+        </nav>
+
+        <!-- Page Size Input -->
+        <div class="abh-page-size-selector">
+          <label for="pageSizeInput">Items per page:</label>
+          <input type="number" id="pageSizeInput" class="jenkins-input" min="1" value="${pageSize}" data-sort-column="${sortColumn}" data-sort-order="${sortOrder}"/>
```
Comment (mawinter69) [inline src/main/resources/io/jenkins/plugins/agent_build_history/AgentBuildHistory/index.jelly:69]: The input field appears too big, I doubt anyone will ever have more than a 3 digit number. So apply some styling via css so the width is limited to 80px

**Comment by mawinter69** [comment]
Let's see if this will really help to solve with the memory problems.

**Comment by meiswjn** [comment]
Nice improvement! Did anyone already try this on a bigger controller and has some feedback?

> Let's see if this will really help to solve with the memory problems.

**Comment by ChristianMa97** [comment]
Yes, we did! We provided feedback on Issue #38. Our Jenkins setup is quite large, with over 2500 jobs and around 300 agents, and everything is running smoothly.

**Pull request merged**